File: subserver/netserver/Core/CommandRegistry.cs
namespace Xrk.Subserver.Core;

public sealed class CommandRegistry
{
    private static readonly Dictionary<string, string> TopAliases = new(StringComparer.Ordinal)
    {
        ["帮助"] = "help",
        ["列表"] = "list",
        ["组"] = "list",
    };

    private static readonly Dictionary<string, string> CmdAliases = new(StringComparer.Ordinal)
    {
        ["状态"] = "status",
        ["更新"] = "update",
        ["同步"] = "sync",
        ["帮助"] = "help",
    };

    private static readonly HashSet<string> ExitWords = new(StringComparer.OrdinalIgnoreCase)
    {
        "exit", "quit", "q", "退出", "离开"
    };

    public static bool IsExitLine(string? line)
    {
        var t = (line ?? "").Trim();
        return t.Length > 0 && ExitWords.Contains(t);
    }

    private static string NormalizeCliLine(string line)
    {
        var parts = line.Trim().Split(' ', StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length == 0) return line.Trim();
        if (TopAliases.TryGetValue(parts[0], out var top)) parts[0] = top;
        else if (parts.Length >= 2 && CmdAliases.TryGetValue(parts[1], out var cmd)) parts[1] = cmd;
        return string.Join(' ', parts);
    }

    public sealed record PluginSet(
        string Group,
        string Description,
        string PluginDir,
        Dictionary<string, Func<IReadOnlyList<string>, Dictionary<string, object?>>> Commands
    );

    private readonly Dictionary<string, PluginSet> _groups = new(StringComparer.OrdinalIgnoreCase);

    public void Register(PluginSet set)
    {
        if (string.IsNullOrWhiteSpace(set.Group)) return;
        _groups[set.Group] = set;
    }

    public IReadOnlyList<string> Groups() => _groups.Keys.OrderBy(x => x).ToList();

    public Dictionary<string, object?> ListHelp()
    {
        var items = Groups().Select(name =>
        {
            var g = _groups[name];
            return new Dictionary<string, object?>
            {
                ["group"] = name,
                ["description"] = g.Description,
                ["commands"] = CommandNames(g)
            };
        }).ToList();

        return new Dictionary<string, object?>
        {
            ["groups"] = items,
            ["count"] = items.Count
        };
    }

    public Dictionary<string, object?> RunLine(string line)
    {
        line = NormalizeCliLine(line);
        if (line.Length == 0)
            return new Dictionary<string, object?> { ["ok"] = false, ["error"] = "空命令" };

        var lower = line.ToLowerInvariant();
        if (lower is "help" or "?")
        {
            var help = ListHelp();
            help["ok"] = true;
            help["hint"] = "用法: <组名> <命令> [参数...]";
            return help;
        }

        if (lower is "list" or "groups")
            return new Dictionary<string, object?> { ["ok"] = true, ["groups"] = Groups() };

        var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        var group = parts[0];
        var cmd = parts.Length > 1 ? parts[1] : "help";
        var args = parts.Length > 2 ? parts.Skip(2).ToList() : Array.Empty<string>();
        return Dispatch(group, cmd, args);
    }

    public Dictionary<string, object?> Dispatch(string group, string cmd, IReadOnlyList<string> args)
    {
        if (!_groups.TryGetValue(group, out var g))
        {
            return new Dictionary<string, object?>
            {
                ["ok"] = false,
                ["error"] = $"未知插件组: {group}",
                ["available"] = Groups()
            };
        }

        cmd = (cmd ?? "help").Trim().ToLowerInvariant();
        if (cmd is "" or "help")
        {
            return new Dictionary<string, object?>
            {
                ["ok"] = true,
                ["group"] = group,
                ["commands"] = CommandNames(g)
            };
        }

        if (cmd == "update")
        {
            var result = PluginKit.DefaultPluginUpdate(g.PluginDir);
            return new Dictionary<string, object?>
            {
                ["ok"] = result["ok"],
                ["group"] = group,
                ["result"] = result
            };
        }

        if (!g.Commands.TryGetValue(cmd, out var handler))
        {
            return new Dictionary<string, object?>
            {
                ["ok"] = false,
                ["error"] = $"未知命令: {cmd}",
                ["group"] = group,
                ["available"] = CommandNames(g)
            };
        }

        var res = new Dictionary<string, object?>(handler(args));
        res.TryAdd("ok", true);
        res["group"] = group;
        return res;
    }

    public List<Dictionary<string, object?>> ApiList()
    {
        return Groups().Select(name =>
        {
            var g = _groups[name];
            return new Dictionary<string, object?>
            {
                ["name"] = name,
                ["description"] = g.Description,
                ["group"] = name
            };
        }).ToList();
    }

    public Dictionary<string, object?> GroupHealth(string group)
    {
        if (!_groups.TryGetValue(group, out var g))
            return new Dictionary<string, object?> { ["ok"] = false, ["error"] = $"未知插件组: {group}" };

        return new Dictionary<string, object?>
        {
            ["ok"] = true,
            ["group"] = g.Group,
            ["name"] = g.Group,
            ["commands"] = CommandNames(g)
        };
    }

    private static List<string> CommandNames(PluginSet g)
    {
        var names = new SortedSet<string>(StringComparer.OrdinalIgnoreCase) { "help", "update" };
        foreach (var key in g.Commands.Keys) names.Add(key);
        return names.ToList();
    }
}

File: subserver/jserver/src/main/java/com/xrk/subserver/core/CommandRegistry.java
package com.xrk.subserver.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.Function;

public final class CommandRegistry {

    private static final Map<String, String> TOP_ALIASES = Map.of(
            "帮助", "help",
            "列表", "list",
            "组", "list"
    );
    private static final Map<String, String> CMD_ALIASES = Map.of(
            "状态", "status",
            "更新", "update",
            "同步", "sync",
            "帮助", "help"
    );

    public static boolean isExitLine(String line) {
        if (line == null) return false;
        String t = line.trim();
        String lower = t.toLowerCase();
        return lower.equals("exit") || lower.equals("quit") || lower.equals("q")
                || t.equals("退出") || t.equals("离开");
    }

    private static String normalizeCliLine(String line) {
        if (line == null) return "";
        String[] parts = line.trim().split("\\s+");
        if (parts.length == 0 || parts[0].isEmpty()) return line.trim();
        if (TOP_ALIASES.containsKey(parts[0])) {
            parts[0] = TOP_ALIASES.get(parts[0]);
        } else if (parts.length >= 2 && CMD_ALIASES.containsKey(parts[1])) {
            parts[1] = CMD_ALIASES.get(parts[1]);
        }
        return String.join(" ", parts);
    }

    public record PluginSet(
            String group,
            String description,
            String pluginDir,
            Map<String, Function<List<String>, Map<String, Object>>> commands
    ) {}

    private final Map<String, PluginSet> groups = new HashMap<>();

    public void register(PluginSet set) {
        if (set.group() == null || set.group().isBlank()) return;
        groups.put(set.group(), set);
    }

    public List<String> groups() {
        return groups.keySet().stream().sorted().toList();
    }

    public Map<String, Object> listHelp() {
        List<Map<String, Object>> items = new ArrayList<>();
        for (String name : groups()) {
            PluginSet g = groups.get(name);
            List<String> cmds = new ArrayList<>(commandNames(g));
            items.add(Map.of(
                    "group", name,
                    "description", g.description() == null ? "" : g.description(),
                    "commands", cmds
            ));
        }
        return Map.of("groups", items, "count", items.size());
    }

    public Map<String, Object> runLine(String line) {
        line = normalizeCliLine(line);
        if (line.isEmpty()) return Map.of("ok", false, "error", "空命令");
        String lower = line.toLowerCase();
        if (lower.equals("help") || lower.equals("?")) {
            Map<String, Object> out = new HashMap<>(listHelp());
            out.put("ok", true);
            out.put("hint", "用法: <组名> <命令> [参数...]");
            return out;
        }
        if (lower.equals("list") || lower.equals("groups")) {
            return Map.of("ok", true, "groups", groups());
        }
        String[] parts = line.split("\\s+");
        String group = parts[0];
        String cmd = parts.length > 1 ? parts[1] : "help";
        List<String> args = parts.length > 2 ? List.of(parts).subList(2, parts.length) : List.of();
        return dispatch(group, cmd, args);
    }

    public Map<String, Object> dispatch(String group, String cmd, List<String> args) {
        PluginSet g = groups.get(group);
        if (g == null) {
            return Map.of("ok", false, "error", "未知插件组: " + group, "available", groups());
        }
        cmd = cmd == null ? "help" : cmd.trim().toLowerCase();
        if (cmd.isEmpty() || cmd.equals("help")) {
            return Map.of("ok", true, "group", group, "commands", commandNames(g));
        }
        if (cmd.equals("update")) {
            Map<String, Object> result = PluginKit.defaultPluginUpdate(g.pluginDir());
            return Map.of("ok", result.get("ok"), "group", group, "result", result);
        }
        var handler = g.commands().get(cmd);
        if (handler == null) {
            return Map.of(
                    "ok", false, "error", "未知命令: " + cmd, "group", group,
                    "available", commandNames(g)
            );
        }
        Map<String, Object> res = new HashMap<>(handler.apply(args == null ? List.of() : args));
        res.putIfAbsent("ok", true);
        res.put("group", group);
        return res;
    }

    public Map<String, Object> apiList() {
        List<Map<String, Object>> out = new ArrayList<>();
        for (String name : groups()) {
            PluginSet g = groups.get(name);
            out.add(Map.of(
                    "name", name,
                    "description", g.description() == null ? "" : g.description(),
                    "group", name
            ));
        }
        return out;
    }

    public Map<String, Object> groupHealth(String group) {
        PluginSet g = groups.get(group);
        if (g == null) {
            return Map.of("ok", false, "error", "未知插件组: " + group);
        }
        return Map.of(
                "ok", true,
                "group", group,
                "name", group,
                "commands", commandNames(g)
        );
    }

    private static List<String> commandNames(PluginSet g) {
        TreeMap<String, Boolean> sorted = new TreeMap<>();
        sorted.put("help", true);
        sorted.put("update", true);
        if (g.commands() != null) g.commands().keySet().forEach(k -> sorted.put(k, true));
        return new ArrayList<>(sorted.keySet());
    }
}
